pyomo/core/expr/expr_pyomo5.py

```diff
@@ -0,0 +1,3434 @@
+        """
+        Compute the polynomial degree of this expression given
+        the degree values of its children.
+
+        This method is called by the :class:`_PolyDegreeVisitor
+        <pyomo.core.expr.current._PolyDegreeVisitor>` class.  It can
+        be over-written by expression classes to customize this
+        logic.
+
+        Args:
+            values (list): A list of values that indicate the degree
+                of the children expression.
+
+        Returns:
+            A nonnegative integer that is the polynomial degree of the
+            expression, or :const:`None`.  Default is :const:`None`.
+        """
+        return None
+
+    def _apply_operation(self, result):     #pragma: no cover
+        """
+        Compute the values of this node given the values of its children.
+
+        This method is called by the :class:`_EvaluationVisitor
+        <pyomo.core.expr.current._EvaluationVisitor>` class.  It must
+        be over-written by expression classes to customize this logic.
+
+        Note:
+            This method applies the logical operation of the
+            operator to the arguments.  It does *not* evaluate
+            the arguments in the process, but assumes that they
+            have been previously evaluated.  But noted that if
+            this class contains auxilliary data (e.g. like the
+            numeric coefficients in the :class:`LinearExpression
+            <pyomo.core.expr.current.LinearExpression>` class, then
+            those values *must* be evaluated as part of this
+            function call.  An uninitialized parameter value
+            encountered during the execution of this method is
+            considered an error.
+
+        Args:
+            values (list): A list of values that indicate the value
+                of the children expressions.
+
+        Returns:
+            A floating point value for this expression.
+        """
+        raise NotImplementedError("Derived expression (%s) failed to "\
+            "implement _apply_operation()" % ( str(self.__class__), ))
+
+
+class NegationExpression(ExpressionBase):
+    """
+    Negation expressions::
+
+        - x
+    """
+
+    __slots__ = ()
+
+    PRECEDENCE = 4
+
+    def nargs(self):
+        return 1
+
+    def getname(self, *args, **kwds):
+        return 'neg'
+
+    def _compute_polynomial_degree(self, result):
+        return result[0]
+
+    def _precedence(self):
+        return NegationExpression.PRECEDENCE
+
+    def _to_string(self, values, verbose, smap, compute_values):
+        if verbose:
+            return "{0}({1})".format(self.getname(), values[0])
+        tmp = values[0]
+        if tmp[0] == '-':
+            i = 1
+            while tmp[i] == ' ':
+                i += 1
+            return tmp[i:]
+        return "- "+tmp
+
+    def _apply_operation(self, result):
+        return -result[0]
+
+
+class NPV_NegationExpression(NegationExpression):
```

Comment:
I still don't understand why we have Net Present Value in our expression system.  We need a better name.